tests/test_main.py

```diff
@@ -900,10 +900,24 @@ class MyModel(BaseModel):
 
 
 def test_parse_root_as_mapping():
-    with pytest.raises(TypeError, match='custom root type cannot allow mapping'):
+    class MyModel(BaseModel):
+        __root__: Mapping[str, str]
 
-        class MyModel(BaseModel):
-            __root__: Mapping[str, str]
+    assert MyModel.parse_obj({1: 2}).__root__ == {'1': '2'}
+
+    with pytest.raises(ValidationError) as exc_info:
+        MyModel.parse_obj({'__root__': {'1': '2'}})
+    assert exc_info.value.errors() == [
+        {'loc': ('__root__', '__root__'), 'msg': 'str type expected', 'type': 'type_error.str'}
+    ]
+
+
+def test_parse_obj_non_mapping_root():
+    class MyModel(BaseModel):
+        __root__: List[str]
+
+    assert MyModel.parse_obj(['a']).__root__ == ['a']
+    assert MyModel.parse_obj({'__root__': ['a']}).__root__ == ['a']
```

Comment:
```suggestion
    assert MyModel.parse_obj({'__root__': ['a']}).__root__ == ['a']
    with pytest.raises(ValidationError, match='...'):
        MyModel.parse_obj({'__not_root__': ['a']})
    with pytest.raises(ValidationError, match='...'):
        MyModel.parse_obj({'__root__': ['a'], 'other': 1})
```

Reply:
Just made a commit with this change. I'm not sure if the `match='...'` was intended to be used verbatim or you meant the `...` as a placeholder. But it worked verbatim, so I committed it like that; let me know if you want a more precise check.

Reply:
sorry :see_no_evil: I was being lazy. `...` as a regex is pretty relaxed!

Reply:
I hadn't seen that syntax before, wasn't sure what exactly it meant. But yeah, I get it now.